Improve minigallery directive path input resolution

lucyleeow opened the pull request:
closes #1356

Uses `relative_to` to get target gallery path - in particular improves the part where we figure out if the file is in a subgallery/subfolder. I think the `relative_to` is more robust and I think it should work as intended.

e.g., 
one glob match: `'.../sphinx-gallery/sphinx_gallery/tests/tinybuild/examples_rst_index/examp_subdir2/plot_sub2.py'`
example dir path: `.../sphinx_gallery/sphinx_gallery/tests/tinybuild/examples_rst_index/`

Uses `resolve` to get abs paths.

cc @story645

lucyleeow commented on sphinx_gallery/directives.py:
```diff
@@ -144,21 +144,41 @@ def has_backrefs(obj):
+                    )
+                # `path` is not inside a `examples_dirs`
+                else:
                     raise ExtensionError(
```
This is the case where the glob match is not inside an `examples_dirs` (i.e., in old code when `len(dirs) = 0`).

lucyleeow commented on sphinx_gallery/directives.py:
```diff
@@ -144,21 +144,41 @@ def has_backrefs(obj):
+                        gal_matches[0][1],
+                    )
+                # `path` inside several `examples_dirs`, take the longest match
+                elif n_match > 1:
+                    ex_parents, target_dir = max(
+                        [(match[0].parents, match[1]) for match in gal_matches],
+                        key=lambda x: len(x[0]),
+                    )
```
I think this may have been what the old `len(dirs)>1` was testing before. The only case where I think this could come about is e.g.,

file: `/a/b/c/d/file.py`
examples dirs (abs path):
* `/a/b/`
* `/a/b/c/d` (i.e. you have a gallery within a sub-sub-folder of another gallery - this *may* be done if someone wanted to use more nesting for their galleries)

In this case, I think it makes sense that we just use the longest matching example dir.

lucyleeow replied:
Hmm I don't think that is the same. E.g., `examples_dirs` in mpl would be e.g., `../galleries/examples`, `../galleries/plot_types/`. These folders have sub-gallaries/sub-folders, but these subfolderes do not need to be added to `examples_dirs` as SG auto checks for 1 level of sub-folders. 

It's only in the case where you want 2 levels of sub-folders e.g., you also have `../galleries/examples/animation/moreanimation/` in your `examples_dirs`.

It would be unusual case. And I don't even know if I can be bothered adding a test for this... :thinking:

lucyleeow reviewed (commented):
I've described how I think we'd get to the errors below.

I am not sure how best to test it, as this would error on a build. I think I could add a test to `test_gen_gallery.py` where I copy over a new `.rst` file that has a minigallery directive that would error (using the `sphinx_app_wrapper.srcdir` to get the right path).

Would this be the best way to do this @larsoner ? I thought about making a new fixture, like `sphinx_app_wrapper` but copying different src files, but I think if I can just copy over files using the same `sphinx_app_wrapper` fixture, it would be better?

larsoner commented:
> Would this be the best way to do this @larsoner ? I thought about making a new fixture, like sphinx_app_wrapper but copying different src files, but I think if I can just copy over files using the same sphinx_app_wrapper fixture, it would be better?

Yeah I think reusing the existing `sphinx_app_wrapper` is reasonable

lucyleeow commented:
This is ready for merge, would be good to get your eye over this @larsoner !

* added tests for the minigallery cases where the file is not in an examples dir and when its from a file in a gallery nested inside another gallery (here the file path matches 2 examples_dirs)
* moved things to use pathlib

lucyleeow commented:
Figuring out `sphinx_app_wrapper` was a wild ride.

For posterity, it was confusing that the `conf_file` fixture and the `conf_file` pytest mark are called the same thing. They are separate and would work fine if they were name differently.

larsoner commented:
> For posterity, it was confusing that the conf_file fixture and the conf_file pytest mark are called the same thing. They are separate and would work fine if they were name differently.

Feel free to open a PR to rename one of them!

larsoner commented:
Thanks @lucyleeow !

lucyleeow commented:
Good idea, somehow I am just hesitant about changing existing 'convention'.
